core/domain/stats_jobs_one_off_test.py

```diff
@@ -1719,3 +1719,66 @@ def test_no_error_when_exploration_is_deleted(self):
             'owner_id', feconf.COMMIT_MESSAGE_EXPLORATION_DELETED)
 
         self.assertItemsEqual(self.run_one_off_job(), [])
+
+    def test_error_when_stats_for_state_is_missing(self):
+        self.save_new_default_exploration('ID', 'owner_id')
+        stats = stats_models.ExplorationStatsModel.get_model('ID', 1)
+        del stats.state_stats_mapping[feconf.DEFAULT_INIT_STATE_NAME]
+        stats.put()
+
+        self.assertItemsEqual(self.run_one_off_job(), [
+            ['UNEXPECTED',
+             'ExplorationStats "ID" v1 does not have stats for card "%s", but '
+             'card appears in version: 1.' % (feconf.DEFAULT_INIT_STATE_NAME)]
+        ])
+
+    def test_error_when_stats_for_state_does_not_exist(self):
+        self.save_new_default_exploration('ID', 'owner_id')
+        stats = stats_models.ExplorationStatsModel.get_model('ID', 1)
+        stats.state_stats_mapping['Unknown State'] = (
+            stats_domain.StateStats.create_default().to_dict())
+        stats.put()
+
+        self.assertItemsEqual(self.run_one_off_job(), [
+            ['UNEXPECTED',
+             'ExplorationStats "ID" v1 has stats for card "Unknown State", but '
+             'card never existed.']
+        ])
+
+    def test_error_when_stats_for_state_no_longer_exists(self):
+        self.save_new_default_exploration('ID', 'owner_id')
+        exp_services.update_exploration('owner_id', 'ID', None, 'noop') # v2
+        stats = stats_models.ExplorationStatsModel.get_model('ID', 2)
+        del stats.state_stats_mapping[feconf.DEFAULT_INIT_STATE_NAME]
+        stats.put()
+
+        self.assertItemsEqual(self.run_one_off_job(), [
+            ['EXPECTED', '1 ExplorationStats model is valid'],
+            ['UNEXPECTED',
+             'ExplorationStats "ID" v2 does not have stats for card "%s", but '
+             'card appears in versions: 1, 2.' % feconf.DEFAULT_INIT_STATE_NAME]
+        ])
+
+    def test_error_when_stats_for_state_once_existed(self):
+        self.save_new_linear_exp_with_state_names_and_interactions(
+            'ID', 'owner_id',
+            ['State ①', 'State ②', 'State ③'],
```

Comment:
Nice, is this to test unicode? If so, I like it :)

Reply:
Yup! Helped catch bugs that have been fixed, too.